diff --git a/dnscrypt-proxy/xtransport.go b/dnscrypt-proxy/xtransport.go
index c9267d4..93bed71 100644
--- a/dnscrypt-proxy/xtransport.go
+++ b/dnscrypt-proxy/xtransport.go
@@ -13,6 +13,7 @@ import (
 	"net"
 	"net/http"
 	"net/url"
+	"syscall"
 	"strconv"
 	"strings"
 	"sync"
@@ -26,6 +27,7 @@ import (
 )
 
 const DefaultFallbackResolver = "9.9.9.9:53"
+const TCP_FASTOPEN_CONNECT = 30
 
 type CachedIPs struct {
 	sync.RWMutex
@@ -99,7 +101,17 @@ func (xTransport *XTransport) rebuildTransport() {
 			}
 			addrStr = ipOnly + ":" + strconv.Itoa(port)
 			if xTransport.proxyDialer == nil {
-				dialer := &net.Dialer{Timeout: timeout, KeepAlive: timeout, DualStack: true}
+				dialer := &net.Dialer{Timeout: timeout, KeepAlive: timeout, DualStack: true,
+					Control: func(network, address string, c syscall.RawConn) error {
+						return c.Control(func(fd uintptr) {
+							err := syscall.SetsockoptInt(int(fd), syscall.SOL_TCP, TCP_FASTOPEN_CONNECT, 1)
+							if err != nil {
+								dlog.Debug("failed to set TCP_FASTOPEN_CONNECT=1")
+								return
+							}
+						})
+					},
+				}
 				return dialer.DialContext(ctx, network, addrStr)
 			} else {
 				return (*xTransport.proxyDialer).Dial(network, addrStr)
